### PR TITLE
fix: corrects and stabilizes finagle instrumentation for http

### PR DESCRIPTION
**Pull request by dmarkwat**
Instruments finagle-http/twitter-util with more rigor than the previous iteration. Chiefly:
- more complete instrumentation of Promise state machine (with inspiration from Kamon)
- instrumentation of the `FuturePool` edge case which otel executors instrumentation can be blind to
- correct chaining of netty http messages to finagle requests across various async boundaries in finagle code paths
- cleanup of previous iterations which were broken in practice but not under test (note: use of different "offload" pools addresses this)

A note on the adjustment/removal/addition of instrumentations: introduction of FuturePool as the "offloader" -- or really any use of FuturePool within the service -- should have been done originally to ensure thread boundaries were crossed and the implementation was sufficiently flexed. Since it wasn't, once a FuturePool was introduced to the previous iteration's tests, the tests broke -- both here in isolation (however thorough) and in practical code applications. The addition of different implementations based on the different threading models (unbounded pool, immediate pool, none defined) is to ensure all different modes are flexed from here on out (Parameterized tests don't work on BeforeAll methods, where the server construction occurs).

### CODE REVIEW

**Comment by dmarkwat** [comment]
Client tests need addressing. Server was the central focus in issue debugging and evaluation. Will mark as ready once complete.

**Comment by dmarkwat** [comment]
Tests are passing, including indy modules. This is now ready for review.

**Comment by dmarkwat** [comment]
Ran some IRL cases, and while I didn't uncover anything concrete, I did notice some edge cases related to scheduling and evaluating the Promise state machine's outcomes that could leave users open to polluted ThreadLocal Contexts. That and a gap in the H2 bridge has now been addressed.

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseInterruptibleInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(isConstructor(), getClass().getName() + "$ConstructorAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class ConstructorAdvice {
+
+    @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
+    @Advice.AssignReturned.ToArguments(@ToArgument(1))
+    public static PartialFunction<Throwable, BoxedUnit> onEnter(
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseInterruptibleInstrumentation.java]: [Javaagent] Method name / annotation mismatch: this method is annotated `@Advice.OnMethodExit` but named `onEnter`. Looks like a copy-paste from the other advices; rename to `onExit` to match the annotation (the repo style explicitly flags this as a copy-paste bug pattern).

```suggestion
    public static PartialFunction<Throwable, BoxedUnit> onExit(
```

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java [hunk 2]
```diff
@@ -99,12 +119,76 @@ protected void initChannel(C channel) throws Exception {
...
+                public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                  // type switch courtesy of
+                  // com.twitter.finagle.netty4.http.Netty4ServerStreamTransport.read()
+                  if (msg instanceof FullHttpRequest) {
+                    FULL_HTTP_REQUEST_CONTEXT.set((FullHttpRequest) msg, Context.current());
+                  } else if (msg instanceof HttpRequest) {
+                    HTTP_REQUEST_CONTEXT.set((HttpRequest) msg, Context.current());
+                  } else {
+                    throw new IllegalArgumentException("unexpected request type: " + msg);
+                  }
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java]: [Javaagent] This `throw` is inside a netty `ChannelInboundHandlerAdapter.channelRead(...)` that we install ourselves into the user's pipeline. Unlike the throws inside advice methods, this is not wrapped by ByteBuddy `suppress = Throwable.class` — it propagates straight back through netty and into the application. Per the javaagent rule "never throw exceptions in javaagent code", drop the throw (the comments above already note this branch shouldn't be reachable). Either log at FINE and continue, or just fall through to `super.channelRead(ctx, msg)`.

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java [hunk 3]
```diff
@@ -99,12 +119,76 @@ protected void initChannel(C channel) throws Exception {
...
+    Context context;
+    // type switch courtesy of com.twitter.finagle.netty4.http.Netty4ServerStreamTransport.read()
+    if (msg instanceof FullHttpRequest) {
+      context = FULL_HTTP_REQUEST_CONTEXT.get((FullHttpRequest) msg);
+    } else if (msg instanceof HttpRequest) {
+      context = HTTP_REQUEST_CONTEXT.get((HttpRequest) msg);
+    } else {
+      // shouldn't practically reach here
+      throw new IllegalArgumentException("unexpected request type: " + msg);
+    }
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java]: [Javaagent] Same pattern: a helper called from advice should not throw. The advice that calls this method has `suppress = Throwable.class`, so the IAE is swallowed at runtime, but it still corrupts the request flow silently for any future caller that doesn't suppress. The comment above already says "shouldn't practically reach here" — prefer returning without setting the context (and optionally logging at FINE) instead of throwing.

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseKInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    public static Scope onApplyEnter(@Advice.This Promise.K thiz) {
+      // if this is null, there's a bug in the instrumentation
+      Context savedContext = TwitterUtilCoreHelpers.PROMISE_K_CONTEXT_FIELD.get(thiz);
+      return savedContext.makeCurrent();
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class, inline = false)
+    public static void onApplyExit(@Advice.Enter Scope scope) {
+      scope.close();
+    }
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseKInstrumentation.java:76]: [Javaagent] `ApplyAdvice` will NPE if `PROMISE_K_CONTEXT_FIELD.get(thiz)` returns `null` (e.g. because `Promise$Transformer` was constructed via a constructor not matched by `TrapContextAdvice`, or the virtual field is not set yet during early class loading). The throw is suppressed by `suppress = Throwable.class`, but it disables instrumentation on that promise and produces noise. Two small fixes:
- mark `onApplyEnter`'s return and the `@Advice.Enter Scope` parameter on `onApplyExit` as `@Nullable`, return `null` when `savedContext == null`, and null-guard `scope.close()` in `onApplyExit`.
- this also matches the pattern already used in `GenStreamingServerDispatcherInstrumentation.DispatchAdvice`.

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/GenStreamingServerDispatcherInstrumentation.java [hunk 5]
```diff
@@ -29,28 +37,29 @@ public ElementMatcher<ClassLoader> classLoaderOptimization() {
...
+    public static Scope methodEnter(@Advice.Argument(0) Object req) {
+      if (req instanceof Request) {
+        // practically this will always be a Request, from HttpServerDispatcher
+        Request request = (Request) req;
+        // if this is null, there's a bug in the instrumentation
+        Context context = request.ctx().apply(Helpers.OTEL_CONTEXT_KEY);
+        return context.makeCurrent();
+      }
+      return null;
     }
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/GenStreamingServerDispatcherInstrumentation.java:58]: [Javaagent] `methodEnter` can return `null` (the `req instanceof Request` branch is taken or `request.ctx().apply(...)` returns `null` when the bridging instrumentation hasn't run for this request). The exit advice already handles a `null` `Scope`, so just add `@Nullable` (and the corresponding import) on the return type here and on the `@Advice.Enter Scope scope` parameter in `methodExit` for accuracy and consistency with the rest of this PR. Also worth null-guarding `context.makeCurrent()` defensively (`return context == null ? null : context.makeCurrent();`) — the comment says it shouldn't be null, but if it ever is, the suppressed NPE silently breaks request propagation.

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java [hunk 6]
```diff
@@ -8,31 +8,48 @@
...
+  private static final VirtualField<HttpRequest, Context> HTTP_REQUEST_CONTEXT =
+      VirtualField.find(HttpRequest.class, Context.class);
 
-  public static Local<Context> contextLocal() {
-    return contextLocal;
-  }
+  public static final RecordSchema.Field<Context> OTEL_CONTEXT_KEY =
+      Request$.MODULE$.Schema().newField();
+
+  public static final String OTEL_NETTY_HANDLER = "otelFinagleNettyHandler";
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java]: [Style] `OTEL_NETTY_HANDLER` is only referenced inside this class (used as the handler name in `mutateHandlerPipeline`). Prefer `private` over `public` for visibility — minimal-necessary-visibility per the style guide.

```suggestion
  private static final String OTEL_NETTY_HANDLER = "otelFinagleNettyHandler";
```

**Review by trask — COMMENTED** [review]
thanks @dmarkwat! AI code review incoming...

The big-picture restructuring (per-Promise-K virtual field, per-FuturePool/respond/transform wrapping, netty→finagle bridging through `Request.ctx()`, and the offloaded vs. immediate test variants) all looks reasonable. A few correctness and javaagent-hygiene issues, mostly around `null`/NPE handling on hot paths and a couple of `throw new IllegalArgumentException(...)` calls in advice helper code that violate the "javaagent code must never throw" rule (they're suppressed inside advice, but the one inside `channelRead` is invoked by netty itself, not by advice, so it is not suppressed and would propagate to the application).

**Comment by dmarkwat** [comment]
@trask , my pleasure -- it was a challenging project, happy to contribute whatever I can! I think I've addressed all the comments and rebased.

File: instrumentation/finagle-http-23.11/javaagent/build.gradle.kts [hunk 7]
```diff
@@ -32,14 +32,23 @@ dependencies {
...
 }
 
+tasks.withType<Jar>().configureEach {
+  // Exclude the Promise stub and its nested classes;
+  // this allows us to compile against these types in the instrumentation
+  // despite them being private in their original inner class;
+  // this is required for VirtualField to have a concrete type to find/get/set on.
+  exclude("com/twitter/util/Promise.class")
+  exclude("com/twitter/util/Promise$*.class")
+}
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/build.gradle.kts]: check out the `compile-stub` pattern for cases like this — see [writing-instrumentation.md](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation.md#instrumenting-code-that-is-not-available-as-a-maven-dependency) (the doc focuses on missing Maven artifacts, but the same mechanism should handle inaccessible/package-private types like `Promise.K`). Existing examples: [`liberty/compile-stub`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/liberty/compile-stub), [`internal-class-loader/compile-stub`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/internal/internal-class-loader/compile-stub).

Suggested:

- Move `Promise.java` into a new sibling module `instrumentation/finagle-http-23.11/compile-stub/` at `src/main/java/com/twitter/util/Promise.java`.
- In `javaagent/build.gradle.kts`, replace the `tasks.withType<Jar>().configureEach { exclude(...) }` block with:
  ```kotlin
  compileOnly(project(":instrumentation:finagle-http-23.11:compile-stub"))

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/TwitterUtilCoreHelpers.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    public InterruptibleWithContext(
+        Context context, PartialFunction<Throwable, BoxedUnit> delegate) {
+      this.context = context;
+      this.delegate = delegate;
+    }
+
+    @Override
+    public boolean isDefinedAt(Throwable x) {
+      try (Scope ignored = context.makeCurrent()) {
+        // Return true only for inputs this function handles`
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/TwitterUtilCoreHelpers.java]:
```suggestion
        // Return true only for inputs this function handles
```

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/TwitterUtilCoreHelpers.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public BoxedUnit apply(Throwable x) {
+      try (Scope ignored = context.makeCurrent()) {
+        return delegate.apply(x);
+      }
+    }
+  }
+
+  public static <T, O> Function1<T, O> wrap(Context context, Function1<T, O> fn) {
+    return (t) -> {
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/TwitterUtilCoreHelpers.java]:
```suggestion
    return t -> {
```

File: testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ContextStorageCloser.java [hunk 10]
```diff
@@ -60,7 +60,10 @@ boolean runWithRestore(AutoCloseable target) {
       } catch (Throwable t) {
         restore();
         if (t instanceof AssertionError) {
-          System.err.println();
+          System.err.println("AssertionError " + t);
+          for (StackTraceElement stackTraceElement : t.getStackTrace()) {
+            System.err.println("\t" + stackTraceElement);
+          }
```
Comment (trask) [inline testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ContextStorageCloser.java:66]: (or, if you think we should keep it let us know why)
```suggestion
          System.err.println();
```

File: instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/FinagleClientExtension.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.UnaryOperator;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * Class-scoped JUnit extension that owns a Finagle {@link Http.Client} per {@link ClientType} and
+ * the underlying {@link EventLoopGroup} for the duration of the test class. Modeled on {@code
+ * Netty41ClientExtension}: build once in {@code beforeAll}, tear down once in {@code afterAll}.
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/FinagleClientExtension.java]:
```suggestion
 * the underlying {@link EventLoopGroup} for the duration of the test class. Clients and services
 * are created lazily on first use and torn down once in {@code afterAll}.
```

File: instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/TwitterUtilCoreInstrumentationTest.java [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+            "parent",
+            () ->
+                pool.apply(
+                    fn0(
+                        () -> {
+                          testing.runWithSpan("pool-child", () -> {});
+                          return 1;
+                        })));
+
+    assertThat((int) Await.result(f, AWAIT)).isEqualTo(1);
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/TwitterUtilCoreInstrumentationTest.java]:
```suggestion
    assertThat(Await.result(f, AWAIT)).isEqualTo(1);
```

File: testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ContextStorageCloser.java (hunk 10 above)
Comment (dmarkwat) [reply testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ContextStorageCloser.java:66]: The problem I encountered with this was that the AssertionError was getting swallowed and so it wasn't especially clear what was being thrown, from where, or why. This just made it apparent what was being thrown so I could trace it back. Seemed more like it was missing/incomplete without it.

File: instrumentation/finagle-http-23.11/javaagent/build.gradle.kts (hunk 7 above)
Comment (dmarkwat) [reply instrumentation/finagle-http-23.11/javaagent/build.gradle.kts]: This is much cleaner. I hadn't seen the pattern before.

**Comment by dmarkwat** [comment]
@trask All comments should be addressed. LMK if there's anything you'd prefer I do with the ContextStorageCloser printing. I found it net helpful when sifting through the leaky storage issues I was experiencing.

File: testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ContextStorageCloser.java (hunk 10 above)
Comment (trask) [reply testing-common/src/main/java/io/opentelemetry/instrumentation/testing/util/ContextStorageCloser.java:66]: Thanks, let's keep it then

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/ChannelTransportInstrumentation.java [hunk 13]
```diff
@@ -25,27 +24,16 @@ public ElementMatcher<TypeDescription> typeMatcher() {
 
   @Override
   public void transform(TypeTransformer transformer) {
-    transformer.applyAdviceToMethod(named("write"), getClass().getName() + "$WriteAdvice");
+    transformer.applyAdviceToMethod(
+        isConstructor(), ChannelTransportInstrumentation.class.getName() + "$ConstructorAdvice");
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/ChannelTransportInstrumentation.java]: [Javaagent] Per the advice-patterns convention, advice references in `transform()` should use `getClass().getName() + "$InnerAdvice"` rather than a `.class` literal — the class literal triggers unwanted class loading in `transform()`.

```suggestion
        isConstructor(), getClass().getName() + "$ConstructorAdvice");
```

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseInterruptibleInstrumentation.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
+    @Advice.AssignReturned.ToArguments(@ToArgument(1))
+    public static PartialFunction<Throwable, BoxedUnit> onExit(
+        @Advice.Argument(1) PartialFunction<Throwable, BoxedUnit> handler) {
+      if (handler instanceof TwitterUtilCoreHelpers.InterruptibleWithContext) {
+        return handler;
+      }
+      Context context = Context.current();
+      return new TwitterUtilCoreHelpers.InterruptibleWithContext(context, handler);
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseInterruptibleInstrumentation.java]: [Style] Throwaway forwarding local — `context` is only used once. Inline `Context.current()` directly into the constructor call.

```suggestion
      return new TwitterUtilCoreHelpers.InterruptibleWithContext(Context.current(), handler);
```

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseKInstrumentation.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  @SuppressWarnings("unused")
+  public static class TrapContextAdvice {
+
+    @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
+    public static void onExit(
+        @Advice.This
+            @SuppressWarnings("rawtypes") // type is from compile-stub and masks private type
+            Promise.K thiz) {
+      Context current = Context.current();
+      TwitterUtilCoreHelpers.PROMISE_K_CONTEXT_FIELD.set(thiz, current);
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/PromiseKInstrumentation.java]: [Style] Throwaway forwarding local — `current` is only used once. Inline `Context.current()` directly into the `set(...)` call.

```suggestion
      TwitterUtilCoreHelpers.PROMISE_K_CONTEXT_FIELD.set(thiz, Context.current());
```

File: instrumentation/finagle-http-23.11/compile-stub/src/main/java/com/twitter/util/Promise.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.twitter.util;
+
+// public accessible stubs of mirrored types in com.twitter.util to ensure compilation;
+// these must be stripped out at jar construction -- see build.gradle.kts
```
Comment (trask) [inline instrumentation/finagle-http-23.11/compile-stub/src/main/java/com/twitter/util/Promise.java]: [General] Comment is incorrect/misleading. The stub module is consumed via `compileOnly(project(":instrumentation:finagle-http-23.11:compile-stub"))` in the javaagent build, so these classes are never on the runtime classpath and never end up in the agent jar — there is no "strip at jar construction" step in `build.gradle.kts`. Suggest rewording to reflect the actual mechanism (compile-only dependency).

**Review by trask — COMMENTED** [review]
Last AI-review round:

**Comment by dmarkwat** [comment]
@trask addressed comments and rebased

File: instrumentation/finagle-http-23.11/javaagent/build.gradle.kts [hunk 17]
```diff
@@ -48,22 +54,44 @@ tasks {
...
+    // ensure concurrent tests are competing for offload pool workers
+    systemProperty("com.twitter.finagle.offload.numWorkers", "2")
+
     systemProperty(
       "metadataConfig",
       "otel.instrumentation.http.client.emit-experimental-telemetry=true," +
         "otel.instrumentation.http.server.emit-experimental-telemetry=true," +
-        "otel.semconv-stability.opt-in=service.peer"
+        "otel.semconv-stability.opt-in=service.peer," +
+        "io.opentelemetry.context.enableStrictContext=true"
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/build.gradle.kts]: only telemetry changing options are needed here

```suggestion
        "otel.semconv-stability.opt-in=service.peer"
```

File: instrumentation/finagle-http-23.11/javaagent/build.gradle.kts [hunk 18]
```diff
@@ -48,22 +54,44 @@ tasks {
...
+    systemProperty("com.twitter.finagle.netty4.numWorkers", "2")
+    // ensure concurrent tests are competing for offload pool workers
+    systemProperty("com.twitter.finagle.offload.numWorkers", "2")
+
     systemProperty(
       "metadataConfig",
       "otel.instrumentation.http.client.emit-experimental-telemetry=true," +
-        "otel.instrumentation.http.server.emit-experimental-telemetry=true"
+        "otel.instrumentation.http.server.emit-experimental-telemetry=true," +
+        "io.opentelemetry.context.enableStrictContext=true"
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/build.gradle.kts]:
```suggestion
        "otel.instrumentation.http.server.emit-experimental-telemetry=true"
```

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/FuturePoolInstrumentation.java [hunk 19]
```diff
@@ -5,43 +5,41 @@
...
 import net.bytebuddy.asm.Advice.AssignReturned.ToArguments.ToArgument;
 import net.bytebuddy.description.type.TypeDescription;
 import net.bytebuddy.matcher.ElementMatcher;
-import scala.Function1;
+import scala.Function0;
 
-class PromiseMonitoredInstrumentation implements TypeInstrumentation {
+/**
+ * Instruments the Promise state machine so that all chains in the Futures/Fibers are
+ * otel-Context-coherent.
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/FuturePoolInstrumentation.java]:
```suggestion
 * Instruments {@link com.twitter.util.ExecutorServiceFuturePool#apply} to wrap the submitted
 * {@link Function0} so it executes under the caller's otel {@link Context}.
```

File: instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java [hunk 20]
```diff
@@ -8,31 +8,48 @@
...
+  public static final RecordSchema.Field<Context> OTEL_CONTEXT_KEY =
+      Request$.MODULE$.Schema().newField();
+
+  private static final String OTEL_NETTY_HANDLER = "otelFinagleNettyHandler";
+
+  private Helpers() {}
 
+  /*
+  Bridges the netty instrumentation to the finagle-netty integration.
+   */
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/Helpers.java]: and the comments below also

```suggestion
  /** Bridges the netty instrumentation to the finagle-netty integration. */
```

File: instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/ServerH2Test.java [hunk 21]
```diff
@@ -57,6 +56,9 @@ private static void assertSwitchingProtocolsEvent(EventDataAssert eventDataAsser
             equalTo(ProtocolSpecificEvent.SWITCHING_PROTOCOLS_TO_KEY, singletonList("h2c")));
   }
 
+  /*
+  Bonus is that this implicitly tests both the server & client h2 upgrades.
+   */
```
Comment (trask) [inline instrumentation/finagle-http-23.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/finaglehttp/v23_11/ServerH2Test.java]:
```suggestion
  /* Bonus is that this implicitly tests both the server & client h2 upgrades. */
```

**Review by trask — APPROVED** [review]
Thanks!

**Comment by dmarkwat** [comment]
@trask done :+1: !
